### PR TITLE
Make replicator executable

### PR DESCRIPTION
**Pull request by Ponewor**
As in the title - this PR attempts to make replicator executable.

We follow the interface from Java and Go versions of the library (at least when it comes to the options we support at the moment).

A few questions:

1. What are the values for `window_size`, `safety_interval` and `sleep_interval` we should use here?
2. Am I assuming correctly that the replicator starts to replicate from the beginning - timestamp 0? This will probably be changed anyway once @macher259 completes progress saving task, right?
3. The `end_timestamp` argument in CDCLogReader looks definitely awkard. IIRC @Gor027 you were working on removing it, right? How is it going?
4. Should we handle SIGINTs gracefully? Is there something to be done here or the SIGINT drops the handle by itself?

### CODE REVIEW

**Comment by piodul** [comment]
> 1. What are the values for `window_size`, `safety_interval` and `sleep_interval` we should use here?

Let's use the defaults from scylla-cdc-go. I believe we have different nomenclature - perhaps we should change it for consistency - but for now, it should be:

- `window_size`, equivalent `QueryTimeWindowSize` (60s)
- `safety_interval`: equivalent `ConfidenceWindowSize` (30s)
- `sleep_interval`: no direct equivalent because scylla-cdc-go uses different delays in case a poll fails, but let's use 10s by default.

It would be nice to be able to change those through command line arguments.

> 2. Am I assuming correctly that the replicator starts to replicate from the beginning - timestamp 0? This will probably be changed anyway once @macher259 completes progress saving task, right?

Yes to both. It should start from the beginning and should resume if the progress was saved.

> 3. The `end_timestamp` argument in CDCLogReader looks definitely awkard. IIRC @Gor027 you were working on removing it, right? How is it going?
> 4. Should we handle SIGINTs gracefully? Is there something to be done here or the SIGINT drops the handle by itself?

AFAIK if you don't register a signal handler, the process is aborted on SIGINT. You can use this: https://docs.rs/tokio/latest/tokio/signal/fn.ctrl_c.html to wait for a ctrlc and handle it gracefully (e.g. let the replicator read up to the current timestamp). After receiving the second ctrlc signal, you can proceed with terminating the process forcefully.

File: scylla-cdc-replicator/src/main.rs [hunk 1]
```diff
@@ -1,29 +1,58 @@
...
+    destination: String,
+
+    #[clap(long, default_value_t = 60)]
+    window_size: u16,
+
+    #[clap(long, default_value_t = 30)]
+    safety_interval: u16,
+
+    #[clap(long, default_value_t = 10)]
+    sleep_interval: u16,
```
Comment (piodul) [inline scylla-cdc-replicator/src/main.rs:31]: Please make it more clear what are the units here. Moreover, if it's seconds then I think a `double` would be a better choice so that sub-second intervals can be specified.

Maybe `clap` allows to just parse `Duration`s?

File: scylla-cdc-replicator/src/main.rs [hunk 2]
```diff
@@ -1,29 +1,58 @@
...
+    keyspace: String,
+
+    #[clap(short, long)]
+    table: String,
+
+    #[clap(short, long)]
+    source: String,
+
+    #[clap(short, long)]
+    destination: String,
```
Comment (piodul) [inline scylla-cdc-replicator/src/main.rs:22]: A short description of the flags that shows on `--help` in the CLI would be appreciated.

File: scylla-cdc-replicator/src/main.rs [hunk 3]
```diff
@@ -1,29 +1,58 @@
...
+        Duration::from_secs(args.safety_interval as u64),
+        Duration::from_secs(args.sleep_interval as u64),
+    )
+    .await?;
+
+    tokio::signal::ctrl_c().await.unwrap();
+
+    println!("Shutting down...");
+
+    replicator.stop();
```
Comment (piodul) [inline scylla-cdc-replicator/src/main.rs:55]: `stop()` does not wait for the replicator to finish, it just sets the upper timestamp so that the log reader stops after reaching it. The whole program will stop immediately if the `main` function returns, regardless of other running tasks.

If you want to gracefully wait until the replicator finishes, you should `.await` then handle returned from the replicator.

**Review by piodul — CHANGES_REQUESTED** [review]

File: scylla-cdc-replicator/src/main.rs (hunk 1 above)
Comment (Ponewor) [reply scylla-cdc-replicator/src/main.rs:31]: Not `Duration`s per se but Duration wrapped in custom type with some traits implemented.
This of course requires parsing command line inputs into durations. Do you have any specific format in mind? Just a finite decimal representation of a number of seconds?

Comment (piodul) [reply scylla-cdc-replicator/src/main.rs:31]: Hmm, no, I didn't have any specific format in mind. I thought that maybe Rust's standard library makes it possible to parse durations in some format, but if not then let's stick to `f64` and interpret as seconds. No need to write any custom wrappers.

File: scylla-cdc/src/log_reader.rs [hunk 4]
```diff
@@ -184,10 +185,10 @@ impl CDCReaderWorker {
...
                         .iter()
-                        .map(|reader| {
+                        .cartesian_product(self.table_names.clone())
+                        .map(|(reader, table_name)| {
                             let reader = Arc::clone(reader);
                             let keyspace = self.keyspace.clone();
-                            let table_name = self.table_name.clone();
                             let factory = Arc::clone(&self.consumer_factory);
                             tokio::spawn(async move {
                                 let consumer = factory.new_consumer().await;
```
Comment (piodul) [inline scylla-cdc/src/log_reader.rs:194]: The consumer should be made aware from which table it consumes data. The new_consumer() method should accept the table/keyspace name.

File: scylla-cdc-replicator/src/main.rs [hunk 5]
```diff
@@ -1,29 +1,65 @@
...
+    let args = Args::parse();
+    let source_tables = args.table.split(',').map(|s| s.to_string()).collect();
+
+    let (mut replicator, handle) = Replicator::new(
+        args.source,
+        args.keyspace.clone(),
+        source_tables,
+        args.destination,
+        args.keyspace,
+        args.table,
```
Comment (piodul) [inline scylla-cdc-replicator/src/main.rs:52]: This doesn't look correct. You adjusted the CDC log reader to be able to read from multiple tables at the same time, but the replicator consumer still supports only one output table. Besides, you are giving it a comma-separated string of table names as the output table, which is guaranteed not to work when more than source table is specified.

Please adjust the replicator so that it supports multiple tables + add a test for it. If you wish, we can postpone it to a followup (but, in such case, please remove the ability to specify more than one source table for the replicator).

**Review by piodul — CHANGES_REQUESTED** [review]

File: scylla-cdc/src/log_reader.rs (hunk 4 above)
Comment (Ponewor) [reply scylla-cdc/src/log_reader.rs:194]: Are you sure you meant the source table? It's not a bad idea but I don't see how it's relevant here.
Maybe you meant the destination table? That would definitely be useful. But when I look at the traits we've defined
```rust
#[async_trait]
pub trait Consumer: Send {
    async fn consume_cdc(&mut self, data: CDCRow<'_>) -> anyhow::Result<()>;
}

#[async_trait]
pub trait ConsumerFactory: Sync + Send {
    async fn new_consumer(&self) -> Box<dyn Consumer>;
}
```

I don't see how a method like
```rust
async fn new_consumer(&self, dest_keyspace: &str, dest_table: &str) -> Box<dyn Consumer>
```
would fit such a general interface. For example `PrinterConsumer` doesn't have a need for destination table/keyspace.

Maybe we could have multiple `ReplicatorConsumerFactories`?

Comment (Ponewor) [reply scylla-cdc/src/log_reader.rs:193]: I've just pushed the implementation with multiple factories for inspection.

Comment (piodul) [reply scylla-cdc/src/log_reader.rs:193]: Sorry for the late response.

The more I think about the multi-table support the less sense I think it makes. I'm not sure there will be much benefit at the moment in supporting it natively in the CDCLogReader vs. just running multiple instances of the CDCLogReader (I just noticed that the scylla-cdc-java replicator does that).

Although running multiple CDCLogReaders duplicates work related to generation fetching, it is a tiny amount of work and complicating the CDCLogReader in order to save work isn't worth it IMO.

Let's drop the multi-table support and just change the replicator to run multiple instances of the CDCLogReader.

Comment (Ponewor) [reply scylla-cdc/src/log_reader.rs:193]: A question then:

How many tables do we expect? Can the handles from `CDCLogReader`s be awaited with something like `join_all` or do we need use `FuturesUnordered` again?

Comment (piodul) [reply scylla-cdc/src/log_reader.rs:193]: I don't think the number of tables will be big, but it is a user-controlled parameter so `FuturesUnordered` won't hurt here.

File: scylla-cdc-replicator/src/main.rs [hunk 6]
```diff
@@ -1,29 +1,68 @@
...
+        Duration::from_secs_f64(args.sleep_interval),
+    )
+    .await?;
+
+    tokio::signal::ctrl_c().await.unwrap();
+
+    println!("Shutting down...");
+
+    replicator.stop();
+    while handles.next().await.is_some() {}
```
Comment (piodul) [inline scylla-cdc-replicator/src/main.rs:66]: I think there is a problem here... If one of the replicators stops due to an error, the whole application won't notice unless the user presses Ctrl+C. The application should react immediately if this happens, i.e. stop other replicators and - ideally - wait until remaining replicators complete.

Moreover, you are not reporting the errors in any way, i.e. no error is returned from main and nothing is printed. Unless there is some logging in the replicator (which I'm not aware of) you won't even know why it stopped and if it was due to a failure or not.

File: scylla-cdc-replicator/src/replicator_consumer.rs [hunk 7]
```diff
@@ -11,7 +11,6 @@ use scylla::prepared_statement::PreparedStatement;
 use scylla::query::Query;
 use scylla::transport::topology::{CollectionType, ColumnKind, CqlType, Table};
 use scylla::Session;
-
```
Comment (piodul) [inline scylla-cdc-replicator/src/replicator_consumer.rs:14]: Unnecessary whitespace change?

**Review by piodul — CHANGES_REQUESTED** [review]

File: scylla-cdc-replicator/src/main.rs [hunk 8]
```diff
@@ -1,29 +1,93 @@
...
+    )
+    .await?;
+
+    loop {
+        select! {
+            _ = tokio::signal::ctrl_c() => {
+                println!("Shutting down...");
+                replicator.stop();
+                while handles.next().await.is_some() {}
+                return Ok(())
```
Comment (piodul) [inline scylla-cdc-replicator/src/main.rs:80]: Nit: missing semicolon at the end of line

File: scylla-cdc-replicator/src/main.rs [hunk 9]
```diff
@@ -1,29 +1,93 @@
...
+            res = handles.next(), if !handles.is_terminated() => {
+                let res = res.unwrap();
+                if res.is_err() {
+                    replicator.stop();
+                    while handles.next().await.is_some() {}
+                    return res;
+                }
+            }
+        }
+        sleep(sleep_interval).await;
```
Comment (piodul) [inline scylla-cdc-replicator/src/main.rs:91]: Why is this sleep needed? It looks like we can arrive here only when `handles.next()` returns a non-error, which happens when one of the CDC log readers finishes reading. Won't it unnecessarily delay shutdown?

File: scylla-cdc-replicator/Cargo.toml [hunk 10]
```diff
@@ -14,4 +14,5 @@ async-trait = "0.1.51"
 anyhow = "1.0.48"
 itertools = "0.10.3"
 uuid = {version = "1.0.0", features = ["v1"]}
-futures-util = "0.3.19"
\ No newline at end of file
+futures-util = "0.3.19"
+log = "0.4.17"
```
Comment (piodul) [inline scylla-cdc-replicator/Cargo.toml:18]: I'd recommend using the [`tracing`](https://docs.rs/tracing/latest/tracing/) crate instead. It integrates with asynchronous code better, and the rust driver uses it, too.

File: scylla-cdc-replicator/src/main.rs (hunk 9 above)
Comment (piodul) [reply scylla-cdc-replicator/src/main.rs:91]: Now that I think about it... Is this code even reachable? The upper timestamp is not set for the log readers, so I don't think they can decide to stop by themselves.

Comment (Ponewor) [reply scylla-cdc-replicator/src/main.rs:91]: You're right

Comment (Ponewor) [reply scylla-cdc-replicator/src/main.rs:91]: In fact the loop itself is useless.

File: scylla-cdc-replicator/src/main.rs [hunk 11]
```diff
@@ -1,29 +1,86 @@
...
+
+    select! {
+        _ = tokio::signal::ctrl_c() => {
+            println!("Shutting down...");
+            replicator.stop();
+            while handles.next().await.is_some() {}
+            Ok(())
+        }
+        res = handles.next() => {
+            replicator.stop();
```
Comment (piodul) [inline scylla-cdc-replicator/src/main.rs:81]: I think we shouldn't stop here unconditionally - we should tell other consumers to stop only if we got an error.

Actually, now that I have wrote the first sentence, I realized that the replicator is supposed to stop only in case of an error. It's not immediately obvious IMO so please add a comment about it.

**Review by piodul — APPROVED** [review]
